test/functional/wallet_migration.py

```diff
@@ -827,6 +831,42 @@ def test_hybrid_pubkey(self):
 
         wallet.unloadwallet()
 
+    def test_failed_migration_cleanup(self):
+        self.log.info("Test that a failed migration is cleaned up")
+        wallet = self.create_legacy_wallet("failed")
+
+        # Make a copy of the wallet with the solvables wallet name so that we are unable
+        # to create the solvables wallet when migrating, thus failing to migrate
+        wallet.unloadwallet()
+        solvables_path = self.nodes[0].wallets_path / "failed_solvables"
+        shutil.copytree(self.nodes[0].wallets_path / "failed", solvables_path)
+        original_shasum = sha256sum_file(solvables_path / "wallet.dat")
+
+        self.nodes[0].loadwallet("failed")
+
+        # Add a multisig so that a solvables wallet is created
+        wallet.addmultisigaddress(2, [wallet.getnewaddress(), get_generate_key().pubkey])
+        wallet.importaddress(get_generate_key().p2pkh_addr)
+
+        assert_raises_rpc_error(-4, "Failed to create new watchonly wallet", wallet.migratewallet)
+
+        assert("failed" in self.nodes[0].listwallets())
+        assert("failed_watchonly" not in self.nodes[0].listwallets())
+        assert("failed_solvables" not in self.nodes[0].listwallets())
+
+        assert(not (self.nodes[0].wallets_path / "failed_watchonly").exists())
+        # Since the file in failed_solvables is one that we put there, migration shouldn't touch it
+        assert(solvables_path.exists())
+        new_shasum = sha256sum_file(solvables_path / "wallet.dat")
+        assert_equal(original_shasum, new_shasum)
+
+        wallet.unloadwallet()
+        with open(self.nodes[0].wallets_path / "failed" / "wallet.dat", "rb") as f:
+            data = f.read(16)
+            _, _, magic = struct.unpack("QII", data)
+            assert_equal(magic, BTREE_MAGIC)
```

Comment:
In bdc828cfa0c:

nit:
I'm assuming this checks if the database is still bdb. Which, if that's the case, a comment indicating it wouldn't hurt.
However, isn't the checksum validation performed earlier already covering this scenario?